### PR TITLE
Added a section on 503 error while accessing headless services

### PR DESCRIPTION
**Pull request by sha-rath**
I observed a few common errors while using StatefulSet with headless services and felt it is good to have it documented under traffic management problems. I have described 3 different scenarios based on the _port name_ specified in the headless service, focusing on the importance of using the Host header while making an HTTP request to a headless service.

[ ] Configuration Infrastructure
[X] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[X] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by ericvn** [comment]
@sha-rath as a new contributor, you may not know how to see the test failure output. In your case the lint_istio.io test failed. If you click on the `Details` link to the right of the test, you will open the job. Under `Build Log` there is a link `Raw build-log.txt` which till take you to the log. There you can see what's wrong. In this case, there are a number of formatting issues with indentation, trailing spaces, etc. You can also run `make lint` locally if you want to see the output without pushing a new commit to GitHub. Locally also works better for spelling errors (you don't have any) in that the misspelled words are highlighted.

**Comment by sha-rath** [comment]
@ericvn Thanks for the explanation. Oops, I missed running make lint before pushing my changes. Just ran make serve to check the appearance. After running make lint locally now, I see the errors and have trouble getting rid of some of them :\ I shall fix and push my changes again. It also detected 2 spelling errors - "behaviour" and "triggerred" on line 338.

**Comment by ericvn** [comment]
@sha-rath Feel free to ask here or on Slack if you need help with things. And lint error free doesn't necessarily mean it looks correct, so doing `make serve` or you can click `Details` on the deploy/netlify task to see what the changes look like is a very good thing to do.

**Comment by sha-rath** [comment]
Thanks for the support. I resolved all the errors except one. 
Errors "Consider starting bulleted lists at the beginning of the line" and "Fenced code blocks should be surrounded by blank lines" are going into a deadlock on lines 334 and 336! (referring to my previous commit here). I see similar usage in other files though.

**Comment by craigbox** [comment]
On behalf of the Commonwealth I wish to apologise that the spell checker doesn't like "behaviour"

**Comment by sha-rath** [comment]
> On behalf of the Commonwealth I wish to apologise that the spell checker doesn't like "behaviour"

No worries, I have changed my "behavior"

**Comment by sha-rath** [comment]
I see that all the tests have passed. Any feedback on the content or do I just have to wait for Merge approval?
Thanks.

**Comment by ericvn** [comment]
@sha-rath At this point you just need to wait on someone to review and approve or get back to you with changes.

**Comment by ericvn** [comment]
@istio/wg-networking-maintainers Any comments on this PR?

**Comment by sha-rath** [comment]
@craigbox Any updates or comments on this one?

**Comment by craigbox** [comment]
We need someone from the networking working group to give it a once-over, and then the docs team will review.  Everyone is heads down with releasing 1.11 at the moment, but we hope to get it looked at soon. Sorry for the delay!

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (sha-rath) [reply content/en/docs/ops/common-problems/network-issues/index.md]: Agree..I removed them.

**Comment by sha-rath** [comment]
@istio/wg-docs-maintainers-english, please take a look at it and merge when possible.

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (rvennam) [inline content/en/docs/ops/common-problems/network-issues/index.md:309]: Needs an explanation here on why it doesn't work with the Host header

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (ericvn) [inline content/en/docs/ops/common-problems/network-issues/index.md:293]: These scripts should be rewritten so they can be cut and pasted into a terminal window. At least sort of. I know you would need to set/find the pod name and IP.  Can you just run the `curl` instead of the `sh`? You will have to provide two for the last bash box, one for each curl.

Comment (rvennam) [reply content/en/docs/ops/common-problems/network-issues/index.md:293]: In other guides, we set 
```export SOURCE_POD=$(kubectl get pod -l app=sleep -o jsonpath='{.items..metadata.name}')```
and then
`kubectl exec -it $SOURCE_POD -c sleep -- sh` or curl directly

**Comment by rvennam** [comment]
I think the PR can benefit from a little more explanation into _why_ it works or doesn't in each scenario. What is it about a headless service that causes the routing differences?

**Comment by sha-rath** [comment]
@rvennam Thanks for going through this in detail. Yes, I get your point but that is all I could figure out from the logs. Therefore, I have only mentioned it in a line or two. Do you have some explanation for this behaviour so that I could add it in as well? I was also actually wondering why does the request not go via the PassthroughCluster in case 1 just like case 2?!

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (sha-rath) [reply content/en/docs/ops/common-problems/network-issues/index.md:293]: Ok, right. I can make these changes. So, I will need to add the SOURCE_POD and also something similar for the pod IP of the nginx sts.

**Comment by rvennam** [comment]
@howardjohn can you shed some light on this behavior?

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (howardjohn) [inline content/en/docs/ops/common-problems/network-issues/index.md:287]: This is a bug, will be fixed in https://github.com/istio/istio/pull/34680

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (howardjohn) [inline content/en/docs/ops/common-problems/network-issues/index.md:289]: nit: "without Host header" is not possible in HTTP/1.1. There is a host header... its just set to the IP (`10.1.1.171`) in this case.

I know what you mean hear, but its not technically 100% accurate

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (howardjohn) [inline content/en/docs/ops/common-problems/network-issues/index.md:306]: A more common way to handle this would probably be to do `curl nginx-0.nginx.default`. Both work, but one is standard way to access specific instance of a headless service via hostname and the other is more or less a workaround to appease Istio.

**Review by howardjohn — COMMENTED** [review]
Would be good to drop in a link to https://istio.io/latest/docs/ops/configuration/traffic-management/traffic-routing/ somewhere. I believe this explains *why* all of these issues occur.

**Comment by rvennam** [comment]
@sha-rath please see john's comments.

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (sha-rath) [reply content/en/docs/ops/common-problems/network-issues/index.md:289]: Yes, the host header was set to 10.1.1.171. So, once istio/istio#34680 is merged in, _curl <POD_IP>_ will work straight away when port name is set to "web" for example (basically auto http headless will work). I'll remove this section then?

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (sha-rath) [reply content/en/docs/ops/common-problems/network-issues/index.md:306]: Good to know that _curl nginx-0.nginx.default_ works fine, the issue is only when pod IP is used. Based on the comment here https://github.com/istio/istio/issues/34679#issuecomment-898561039, I understand that when the port name is explicitly set to "http", the problem described in my 2nd section will **still** exist. However, if using pod IP is not the right way to access a headless service, should I rephrase or remove the 2nd section (with port name set to http)?

**Comment by sha-rath** [comment]
Thanks for the reminder @rvennam. Any thoughts on how I could make this doc better? Or does this become trivial once your fix is implemented and also considering a more common way of using the hostname to access an instance of a headless service? @howardjohn

**Comment by linsun** [comment]
@sha-rath, thank you for taking the initiative to document this for everyone!!

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (frankbu) [inline content/en/docs/ops/common-problems/network-issues/index.md:292]: This doesn't display properly in preview and won't work with cut-paste. How about calling curl directly?
```suggestion
$ kubectl exec -it $SOURCE_POD -c sleep -- curl 10.1.1.171 -s -o /dev/null -w "%{http_code}"
503
```

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (frankbu) [inline content/en/docs/ops/common-problems/network-issues/index.md:307]: See previous comment.
```suggestion
    $ kubectl exec -it $SOURCE_POD -c sleep -- curl -H "Host: nginx.default" 10.1.1.171 -s -o /dev/null -w "%{http_code}"
    200
```

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (frankbu) [inline content/en/docs/ops/common-problems/network-issues/index.md:324]:
```suggestion
    $ kubectl exec -it $SOURCE_POD -c sleep -- curl 10.1.1.171 -s -o /dev/null -w "%{http_code}"
    200
    {{ /text >}}
    
    {{< text bash >}}
    $ kubectl exec -it $SOURCE_POD -c sleep -- curl -H "Host: nginx.default" 10.1.1.171 -s -o /dev/null -w "%{http_code}"
    200
```

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (frankbu) [inline content/en/docs/ops/common-problems/network-issues/index.md:335]:
```suggestion
    $ kubectl exec -it $SOURCE_POD -c sleep -- curl web-0.nginx.default -s -o /dev/null -w "%{http_code}"
    200
```

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (sha-rath) [reply content/en/docs/ops/common-problems/network-issues/index.md:292]: I thought about this too, but the problem is Pod IP (10.1.1.171 here) is not a constant. When this command is copy-pasted from the doc page, wrong Pod IP would be copied always and the user will have to then manually change it. If you keep it the way it is now, copy-paste will exec into the sleep container and wait for the user's curl <Pod IP> input.

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (sha-rath) [reply content/en/docs/ops/common-problems/network-issues/index.md:324]: Same problem as mentioned above

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (sha-rath) [reply content/en/docs/ops/common-problems/network-issues/index.md:335]: In my example, web-0 is a constant. So, this will work and could be changed.

File: content/en/docs/ops/common-problems/network-issues/index.md
Comment (frankbu) [reply content/en/docs/ops/common-problems/network-issues/index.md:292]: Yes, but the problem with the way you have it is that the copy button doesn't copy any of the curl command (and it doesn't display so nicely either - see [preview](https://deploy-preview-10071--preliminary-istio.netlify.app/latest/docs/ops/common-problems/network-issues/#503-error-while-accessing-headless-services)) so the user needs to type the whole line. In other similar docs the copy will pick up the whole command so that after the paste, the user can just hit the `back-arrow` key and edit the IP before hitting `enter`.

In other similar docs we either mention that `10,.1.1.171` is the IP of one of the pods or we document it as `curl <pod ip> -s -o /dev/null -w "%{http_code}"` and tell the user to change `<pod ip>` to the IP of one of the pods.

Comment (sha-rath) [reply content/en/docs/ops/common-problems/network-issues/index.md]: Thanks for the explanation!
I have pushed all these changes in a new commit.

**Review by frankbu — APPROVED** [review]

**Pull request merged**
